components/eventing-controller/controllers/subscriptionv2/eventmesh/test/reconciler_integration_test.go

```diff
@@ -0,0 +1,426 @@
+func TestMain(m *testing.M) {
+	// Note: The setup will provision a single K8s env and
+	// all the tests need to create and use a separate namespace
+
+	// setup env test
+	if err := setupSuite(); err != nil {
+		panic(err)
+	}
+
+	// run tests
+	code := m.Run()
+
+	// tear down test env
+	if err := tearDownSuite(); err != nil {
+		panic(err)
+	}
+
+	os.Exit(code)
+}
+
+func Test_CreateSubscription(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name                     string
+		givenSubscriptionFunc    func(namespace string) *eventingv1alpha2.Subscription
+		wantSubscriptionMatchers gomegatypes.GomegaMatcher
+	}{
+		{
+			name: "should fail to create subscription if types is empty",
+			givenSubscriptionFunc: func(namespace string) *eventingv1alpha2.Subscription {
+				return reconcilertesting.NewSubscription("test", namespace,
+					reconcilertesting.WithDefaultSource(),
+					reconcilertesting.WithEmptyTypes(),
+					reconcilertesting.WithWebhookAuthForBEB(),
+					reconcilertesting.WithSinkURL(reconcilertesting.ValidSinkURL(namespace, "test")),
+				)
+			},
+			wantSubscriptionMatchers: gomega.And(
+				reconcilertesting.HaveSubscriptionNotReady(),
+				reconcilertesting.HaveCondition(eventingv1alpha2.MakeCondition(
+					eventingv1alpha2.ConditionSubscribed,
+					eventingv1alpha2.ConditionReasonSubscriptionCreationFailed,
+					corev1.ConditionFalse, "Types are required")),
+				reconcilertesting.HaveCleanEventTypesEmpty(),
+			),
+		},
+		{
+			name: "should succeed to create subscription if types are non-empty",
+			givenSubscriptionFunc: func(namespace string) *eventingv1alpha2.Subscription {
+				return reconcilertesting.NewSubscription("test", namespace,
+					reconcilertesting.WithDefaultSource(),
+					reconcilertesting.WithTypes([]string{
+						fmt.Sprintf("%s0", reconcilertesting.OrderCreatedV1EventNotClean),
+						fmt.Sprintf("%s1", reconcilertesting.OrderCreatedV1EventNotClean),
+					}),
+					reconcilertesting.WithWebhookAuthForBEB(),
+					reconcilertesting.WithSinkURL(reconcilertesting.ValidSinkURL(namespace, "test")),
+				)
+			},
+			wantSubscriptionMatchers: gomega.And(
+				reconcilertesting.HaveSubscriptionActiveCondition(),
+				reconcilertesting.HaveCleanEventTypes([]eventingv1alpha2.EventType{
+					{
+						OriginalType: fmt.Sprintf("%s0", reconcilertesting.OrderCreatedV1EventNotClean),
+						CleanType:    fmt.Sprintf("%s0", reconcilertesting.OrderCreatedV1Event),
+					}, {
+						OriginalType: fmt.Sprintf("%s1", reconcilertesting.OrderCreatedV1EventNotClean),
+						CleanType:    fmt.Sprintf("%s1", reconcilertesting.OrderCreatedV1Event),
+					},
+				}),
+			),
+		},
+	}
+
+	for _, testCase := range testCases {
+		tc := testCase
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			g := gomega.NewGomegaWithT(t)
+			ctx := context.Background()
+
+			// create unique namespace for this test run
+			testNamespace := getTestNamespace()
```

Comment:
Just an idea. As an alternative, you could also have defined an int with value 0 at line 45 and then, at this line, build a namespace out of it and `++` it. But I guess this works just as well.

Reply:
The tests are running with `t.parallel`. I think using a common variable could result in race condition.